### PR TITLE
add exchange FTX

### PR DESCRIPTION
**Pull request by RuChiang**
Hi @kroitor,

Please kindly assist me to merge this integration for the FTX exchange.
Since the exchange is launching some new trading programs, and might need this quite urgently, it'll be much appreciated if you can help look into this one.
Let me know if you have any issues!

Many thanks!

### CODE REVIEW

**Comment by kroitor** [comment]
Related: https://github.com/ccxt/ccxt/issues/5643

**Comment by RuChiang** [comment]
Hi @kroitor 

I have actually already run the linter and the npm run build, and they all worked fine on my side.
I just checked the log on the auto-build above, seems like its function parameters issue?

```
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-contribute-code
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes
PHP Fatal error:  Uncaught ErrorException: Declaration of ccxt\ftx::handle_errors($code, $reason, $url, $method, $headers, $body, $response) should be compatible with ccxt\Exchange::handle_errors($code, $reason, $url, $method, $headers, $body, $response, $request_headers, $request_body) in /home/travis/build/ccxt/ccxt/php/ftx.php on line 620 in /home/travis/build/ccxt/ccxt/php/test/syntax.php:12
Stack trace:
#0 /home/travis/build/ccxt/ccxt/ccxt.php(59): on_error(2, 'Declaration of ...', '/home/travis/bu...', 620, Array)
#1 /home/travis/build/ccxt/ccxt/ccxt.php(59): require_once()
#2 [internal function]: ccxt\{closure}('ftx')
#3 /home/travis/build/ccxt/ccxt/php/test/syntax.php(21): spl_autoload_call('ccxt\\ftx')
#4 {main}
  thrown in /home/travis/build/ccxt/ccxt/php/test/syntax.php on line 12
```

**Comment by RuChiang** [comment]
Now I actually realized that the parameter set of ```handle_errors``` function in exchange.php is different from ```handleErrors``` function in Exchange.js, I wonder if I shall just modified the handle_errors function locally and submit the commits. But I guess this will possibly tamper the build transpiling process?

**Comment by kroitor** [comment]
>  I wonder if I shall just modified the handle_errors function locally and submit the commits. But I guess this will possibly tamper the build transpiling process?

Nope, no need to modify the base classes, just copy the function signature – the first line of that `handleErrors` function (the name and the list of arguments) from any other exchange.

> seems like its function parameters issue?

There are other issues throughout the file, and there's many of them, namely, we need to make sure that we follow the guidelines from here precisely to a word:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

For example, the `ftx.js` file voilates this rule frequently:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys

And, many other rules outlines there. Unfortunately, violating any of them will break the build or will not work with this or that language. So, we have to be extremely pedantic in punctuation and code. The best approach is to take each line of `ftx.js` file and check if it violates any of the rules, then take next line and so on. Also a good practice would be to copy or look up how things are done in the other exchanges, especially, big and popular ones.

**Comment by RuChiang** [comment]
Thanks for your kind reply pointing out the out-standing issues.
I just realized my branch wasn't update to date, and that's why I got an outdated function signature.
I will try to fix them ASAP.
By the way, is there any possible way to run the `test` from travis.build locally?

**Comment by RuChiang** [comment]
@kroitor 

I think I have resolved most of the issues, but travis build now fails without explaining any reasons, as shown below.
```
FAILED ftx (Python): (explain @ run-tests.js:187)
	ftx using proxy `` (explain @ run-tests.js:190)
```
Would you assist me to find what the actual problem is?

 Many thanks for your help!!!

**Comment by RuChiang** [comment]
Hi @kroitor 
I have updated the code according to the requirements, and also has passed the build. Could you have a look?

Many thanks

**Comment by RuChiang** [comment]
Hi @kroitor ,
Would you kindly let me know how long the code review is going to take approximately?

Many thanks!!

**Comment by kroitor** [comment]
@RuChiang it still does violate many of the CONTRIBUTING rules and has some errors, that we will have to fix first. There's very little point in waiting for a review – if it's not merged, that means it has obvious problems.

For example:

----
1. violating the .toString rule:
-  https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R122
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes
----
2. duplicate lines + undefined return values:
- https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R292
- this check should be in `handleErrors` instead of each other method
----
3. dict keys access violations (all over the file):
- https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R467
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys
----
4. this code should be in parseTicker:
- https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R201
----
5. Not using safe methods here:
- https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R253
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sanitizing-input-with-safe-methods
----
6. default `{}` value of `safeValue` here is not for-iterable on the next line:
- https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R338
----
7. etc...
----
So, basically, the code still has to be pedantically edited line by line to match the CONTRIBUTING guides.

**Comment by RuChiang** [comment]
Hi @kroitor 
Thanks for pointing them out. I'm on it fixing now.
I'm still a little confused with this __dict keys access violations__. The CONTRIBUTING file says we should use __single-quoted string key notation__. May I know what's the problem with writing ```response['success']``` then? As you've metioned
> 3. dict keys access violations (all over the file):
> https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R467
> https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys


Many thanks!

**Comment by kroitor** [comment]
> May I know what's the problem with writing response['success'] then?

The answer is explained by that linked posted there: 

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys )

Let me know if that does not answer the question. It isn't about quotes, really, but more about keys that might not be there when you access them. JS tolerates that, but other languages don't.

**Comment by RuChiang** [comment]
Hi @kroitor ,

I have again updated the file. Please kindly check it if you have time. Really appreciate it!

Also let me know if there is anything else that needs to be changed.

**Comment by RuChiang** [comment]
Hi @kroitor ,

Any updates on this?? Big thanks

**Comment by kroitor** [comment]
@RuChiang I will review asap and will post comments along the lines of code, there's still many inconsistencies and redundant lines, namely:

1. `fetchMarkets` should fetch futures and markets and we will need to add a `type` field to the market structure to differentiate between the two types of instruments, see how that is done in [okex3](https://github.com/ccxt/ccxt/blob/master/js/okex3.js):

    - https://github.com/ccxt/ccxt/blob/master/js/okex3.js#L473
    - https://github.com/ccxt/ccxt/blob/master/js/okex3.js#L491
    - https://github.com/ccxt/ccxt/blob/master/js/okex3.js#L615

    If you have a difficulty with it, I'll edit and fix that as soon as I can.

2. The `parseTrade` method violates [this rule](https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#parsing-symbols), 

3. `fetchOrderBook` should return `this.parseOrderBook` (see how it is implemented in most other exchanges)

4. `fetchTicker` has redundant calls to `fetchTicker` (see how we use variable method names to switch endpoints [here](https://github.com/ccxt/ccxt/blob/master/js/binance.js#L423), [here](https://github.com/ccxt/ccxt/blob/master/js/bitbay.js#L986) or [here](https://github.com/ccxt/ccxt/blob/master/js/bitfinex2.js#L596) and in many other examples where a `method` variable is introduced):

```JavaScript
    async fetchTicker (symbol, params = {}) {
        await this.loadMarkets ();
        const market = this.market (symbol);
        const response = {};
        const marketType = this.safeString (market['info'], 'type');
        if (marketType !== undefined && marketType === 'future') {
            const request = {};
            request['market'] = market['id'];
            const future = await this.publicGetFuturesMarket (this.extend (request, params));
            return this.parseTicker (this.extend (response, future['result']), market);
        }
        return this.parseTicker (response, market);
    }
```

↓

```JavaScript
    async fetchTicker (symbol, params = {}) {
        await this.loadMarkets ();
        const market = this.market (symbol);
        const request = {
            'market': market['id'],
        };
        // market['type'] must be populated in `fetchMarkets`
        if (market['type'] !== 'future') {
            throw new NotSupported (this.id + ' fetchTicker() supports futures markets only, see exchange.markets[symbol]["type"] for details');
        }
        const response = = await this.publicGetFuturesMarket (this.extend (request, params));
        const result = this.safeValue (response, 'result', {})
        return this.parseTicker (result, market);
    }
```

5. The `timestamp` parsing should be in `parseTrade`, not in `fetchTrades`: https://github.com/ccxt/ccxt/pull/5807/files#diff-fcf6bf727d557368937548bcda026434R317

6. Current implementation of `createOrder` is redundant and has too much duplicated code, apart from several logical errors, it does not set the `type` for `stop` conditional orders:

```JavaScript
    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
        // type can be a market, limit, or stop-loss
        await this.loadMarkets ();
        const market = this.market (symbol);
        if (side !== 'buy' && side !== 'sell') { // check side value
            throw new InvalidOrder ('Invalid order side value');
        }
        if (type === 'market' || type === 'limit') { // check order type
            const request = {
                'market': market['id'],
                'side': side,
                'price': price,
                'type': type,
                'size': parseFloat (this.amountToPrecision (symbol, amount)),
            };
            const response = await this.privatePostOrders (this.extend (request, params));
            const result = this.safeValue (response, 'result', {});
            return this.parseOrder (result);
        } else if (type === 'stop') {
            const request = {
                'market': this.marketId (symbol),
                'side': side,
                'triggerPrice': price,
                'size': parseFloat (this.amountToPrecision (symbol, amount)),
            };
            const response = await this.privatePostConditionalOrders (this.extend (request, params));
            const result = this.safeValue (response, 'result', {});
            return this.parseOrder (result);
        } else {
            throw new InvalidOrder ('Invalid order type');
        }
    }
```

↓

```JavaScript
    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
        // type can be a market, limit, or stop-loss
        await this.loadMarkets ();
        const market = this.market (symbol);
        const request = {
            'market': market['id'],
            'side': side,
            'type': type,
            'size': amount,
        }
        let method = 'privatePostConditionalOrders';
        if ((type === 'limit') || (type === 'market')) {
            method = 'privatePostOrders';
            request['price'] = price;
        } else if ((type === 'stop') || (type === 'takeProfit')) {
            request['triggerPrice'] = price;
        } else if (type === 'trailingStop') {
            request['trailValue'] = price;
        } else {
            throw new InvalidOrder (this.id + ' createOrder () does not support order type ' + type + ', supported types: limit, market, stop, trailingStop, takeProfit');
        }
        const response = await this[method] (this.extend (request, params));
        const result = this.safeValue (response, 'result', {});
        return this.parseOrder (result, market);
    }
```

We cannot use `amountToPrecision` there, because market precision is set to undefined in `fetchMarkets`.

7. We have to return ISO8601 strings in the `datetime` field in UTC in all parsers, `parseTicker()`, `parseTrade()`, `parseOrder()`, etc – for that we need to use `this.iso8601 (timestamp)` to format the string to UTC.

8. I will post more comments on other issues, the review in progress

**Comment by RuChiang** [comment]
Hi @kroitor ,
Thanks for pointing out the issues! I have made some fixes based on your suggestions.
But I have some points that I'd like to discuss
> The parseTrade method violates [this rule](https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#parsing-symbols),
I have checked the example that you provided. The problem here is that there's neither `id` nor `symbol` returned from the FTX API endpoint, and therefore the only way I can return `id` and `symbol` is to check the info in the market object that's passed into `parseTrade`
> We have to return ISO8601 strings in the datetime field in UTC in all parsers, parseTicker(), parseTrade(), parseOrder(), etc – for that we need to use this.iso8601 (timestamp) to format the string to UTC.
FTX only returns datetime string, and therefore I think only `this.parse8601()` shall be needed? The datetime can be directly assigned with the value from `this.safeString(trade, 'time')`
> Current implementation of createOrder is redundant and has too much duplicated code, apart from several logical errors, it does not set the type for stop conditional orders
Apparently FTX updated the endpoints :stuck_out_tongue: Just realized this.
Please let me know if there are anything I can do/update to speed up the review process. Big thanks!!

**Comment by RuChiang** [comment]
Hi @kroitor ,
Thanks for pointing out the issues! I have made some fixes based on your suggestions.
But I have some points that I'd like to discuss
> The parseTrade method violates [this rule](https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#parsing-symbols),

I have checked the example that you provided. The problem here is that there's neither `id` nor `symbol` returned from the FTX API endpoint, and therefore the only way I can return `id` and `symbol` is to check the info in the market object that's passed into `parseTrade`
> We have to return ISO8601 strings in the datetime field in UTC in all parsers, parseTicker(), parseTrade(), parseOrder(), etc – for that we need to use this.iso8601 (timestamp) to format the string to UTC.

FTX only returns datetime string, and therefore I think only `this.parse8601()` shall be needed? The datetime can be directly assigned with the value from `this.safeString(trade, 'time')`
> Current implementation of createOrder is redundant and has too much duplicated code, apart from several logical errors, it does not set the type for stop conditional orders

Apparently FTX updated the endpoints :stuck_out_tongue: Just realized this.
Also I have added FTX into `key.json` to skip `run-test` as throwing NotSupported Error from `fetchTicker` will fail the build.
Please let me know if there are anything I can do/update to speed up the review process. Big thanks!!

**Comment by RuChiang** [comment]
Hi @kroitor ,
any updates?? Many thanks

**Comment by kroitor** [comment]
@RuChiang thanks for your patience! I'll do my best to get to it as soon as I can, will post here, hopefully very soon.

**Comment by kelonye** [comment]
👍

**Comment by kroitor** [comment]
This PR will be merged within 10 days or sooner. Thx for your patience!
